ui/src/pages/AllStories.tsx

```diff
@@ -21,111 +85,247 @@ function createData(
+            story.year,
+            story.is_visible,
+            story.author_first_name,
+            story.author_last_name,
+            story.author_country,
+            story.image_url,
+            story.video_url,
+            story.content
+          );
+        })
+      );
 
-  const [clickedStory, setClickedStory] = useState<Story | undefined>(
+  const { data: allStories, error } = useSWR<StoryView[] | undefined>(
+    "/api/stories",
+    fetchStories
+  );
+  const [clickedStory, setClickedStory] = useState<StoryView | undefined>(
     undefined
   );
+  const classes = useStyles();
+
+  useEffect(() => {
+    dispatch({ type: "INITIALIZE_AFTER_API", rows: allStories });
+  }, [allStories]);
 
   const setClickedRow = (rowId: number | undefined) => {
     if (rowId) {
-      const story = allStories?.find((story) => story.ID === rowId);
+      const story = allStories?.find((story: StoryView) => story.ID === rowId);
       if (story) {
         setClickedStory(story);
       }
     }
   };
 
-  const handleChange = (e, d) => {
-    if (e.target.checked) {
-      setVisibleState((prevState) => [...prevState, d.id]);
-    } else {
-      setVisibleState((prevState) => prevState.filter((e) => e !== d.id));
-    }
+  const handleSwitchChange = (e, story) => {
+    dispatch({ type: "HANDLE_SWITCH_CHANGE", e, story });
+    mutate(
+      "/api/stories",
+      (prevStories: Story[]) => {
+        return prevStories.map((currStory) =>
+          currStory.ID === story.ID
+            ? { ...currStory, is_visible: !currStory.is_visible }
+            : currStory
+        );
+      },
+      false
+    );
   };
 
   const handleCheckedAll = () => {
-    setSelectedRowIds((prevState) =>
-      prevState.length === rows.length ? [] : rows.map((d) => d.id)
-    );
+    dispatch({ type: "HANDLE_CHECKED_ALL", rows: allStories });
   };
-  const handleChecked = (e, d) => {
-    if (e.target.checked) {
-      setSelectedRowIds((prevState) => [...prevState, d.id]);
-    } else {
-      setSelectedRowIds((prevState) => prevState.filter((e) => e !== d.id));
-    }
+  const handleChecked = (e, story) => {
+    dispatch({ type: "HANDLE_CHECKED", e, story });
   };
 
-  const handleRequestSort = (property) => {
-    const isDesc = orderBy === property && order === "asc";
-    const newOrder = isDesc ? "desc" : "asc";
-    setOrder(newOrder);
-    setOrderBy(property);
+  const handleRequestSort = (property: string) => {
+    const isDesc = state.orderBy === property && state.order === "asc";
+    const order = isDesc ? "desc" : "asc";
+    dispatch({ type: "SET_ORDERING", order, orderBy: property });
 
-    onChangeTableSort(newOrder, property);
+    onChangeTableSort(order, property);
   };
   const onChangeTableSort = (order, property) => {
-    const newData = stableSort(tableData, getComparator(order, property));
-    setTableData(newData);
+    const data = stableSort(state.tableData, getComparator(order, property));
```

Comment:
Can we move the `stableSort` function out of the component? We can put it into a `helpers.ts` file.

Reply:
hmm I think it's unnecessary just for the one function? I think it's better to just keep it with the other `tableSort` functions